File: examples/agents/typescript/agent.ts
declare const process: {
  env: Record<string, string | undefined>;
  exitCode: number;
};

type Actor = { id: string };
type Checkpoint = { last_event_seq: number };
type Message = {
  id: string;
  chat_id: string;
  actor_id: string;
  thread_root_id?: string | null;
  mentioned_actor_ids: string[];
};
type Frame = {
  op?: string;
  seq?: number;
  type?: string;
  current_seq?: number;
  data?: unknown;
  code?: string;
  fatal?: boolean;
};

const coreURL = required("COMA_CORE_URL").replace(/\/$/, "");
const apiKey = required("COMA_AGENT_API_KEY");
const consumer = process.env.COMA_RUNTIME_CONSUMER ?? "external-ts-example";
const websocketURL = `${coreURL.replace(/^http/, "ws")}/api/v1/ws`;
let actorID = "";

async function request<T>(
  path: string,
  method = "GET",
  body?: unknown,
): Promise<T> {
  const response = await fetch(`${coreURL}${path}`, {
    method,
    headers: {
      Authorization: `Bearer ${apiKey}`,
      ...(body === undefined ? {} : { "Content-Type": "application/json" }),
    },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  if (!response.ok) {
    throw new Error(
      `${method} ${path}: ${response.status} ${await response.text()}`,
    );
  }
  return (await response.json()) as T;
}

async function connectOnce(): Promise<void> {
  const checkpoint = await request<Checkpoint>(
    `/api/v1/agent-runtime/checkpoints/${encodeURIComponent(consumer)}`,
  );
  const socket = new WebSocket(websocketURL);
  let chain = Promise.resolve();

  await new Promise<void>((resolve, reject) => {
    let settled = false;
    const fail = (cause: unknown) => {
      if (settled) return;
      settled = true;
      socket.close();
      reject(cause);
    };

    socket.addEventListener("open", () => {
      socket.send(
        JSON.stringify({
          op: "auth",
          request_id: crypto.randomUUID(),
          access_token: apiKey,
          last_seq: checkpoint.last_event_seq,
        }),
      );
    });
    socket.addEventListener("message", (event) => {
      chain = chain
        .then(() =>
          handleFrame(socket, JSON.parse(String(event.data)) as Frame),
        )
        .catch(fail);
    });
    socket.addEventListener("error", () => fail(new Error("WebSocket error")));
    socket.addEventListener("close", () => {
      chain.then(() => {
        if (settled) return;
        settled = true;
        resolve();
      }, fail);
    });
  });
}

async function handleFrame(socket: WebSocket, frame: Frame): Promise<void> {
  if (frame.op === "hello") {
    console.log(`Connected as ${actorID}; resume is active.`);
    return;
  }
  if (frame.op === "resync_required") {
    await saveCheckpoint(requiredNumber(frame.current_seq, "current_seq"));
    socket.close(1000, "checkpoint resynchronized");
    return;
  }
  if (frame.op === "error") {
    console.error(`WebSocket error frame: ${frame.code ?? "unknown"}`);
    if (frame.fatal) socket.close();
    return;
  }
  if (frame.op !== "event") return;

  const seq = requiredNumber(frame.seq, "seq");
  if (frame.type === "message.created" && isMessage(frame.data)) {
    const message = frame.data;
    if (
      message.actor_id !== actorID &&
      message.mentioned_actor_ids.includes(actorID)
    ) {
      await reply(message, seq);
      console.log(`Replied to mention ${message.id} at event ${seq}.`);
    }
  }
  await saveCheckpoint(seq);
  if (socket.readyState === WebSocket.OPEN) {
    socket.send(JSON.stringify({ op: "ack", seq }));
  }
}

async function reply(message: Message, seq: number): Promise<void> {
  await request(`/api/v1/agent-tools/reply_in_thread`, "POST", {
    arguments: {
      chat_id: message.chat_id,
      client_msg_id: await stableUUID(`reply:${seq}`),
      body: `I received your mention (message \`${message.id}\`).`,
      body_format: "markdown",
      thread_root_id: message.thread_root_id ?? message.id,
      mentioned_actor_ids: [],
      file_ids: [],
    },
    correlation_id: await stableUUID(`correlation:${seq}`),
    confirmed: true,
  });
}

async function saveCheckpoint(seq: number): Promise<void> {
  await request(
    `/api/v1/agent-runtime/checkpoints/${encodeURIComponent(consumer)}`,
    "PUT",
    { last_event_seq: seq },
  );
}

async function stableUUID(label: string): Promise<string> {
  const digest = new Uint8Array(
    await crypto.subtle.digest(
      "SHA-256",
      new TextEncoder().encode(`${actorID}:${label}`),
    ),
  ).slice(0, 16);
  digest[6] = (digest[6]! & 0x0f) | 0x50;
  digest[8] = (digest[8]! & 0x3f) | 0x80;
  const hex = [...digest]
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

function isMessage(value: unknown): value is Message {
  if (!value || typeof value !== "object") return false;
  const message = value as Partial<Message>;
  return (
    typeof message.id === "string" &&
    typeof message.chat_id === "string" &&
    typeof message.actor_id === "string" &&
    Array.isArray(message.mentioned_actor_ids)
  );
}

function required(name: string): string {
  const value = process.env[name]?.trim();
  if (!value) throw new Error(`Missing environment variable ${name}`);
  return value;
}

function requiredNumber(value: unknown, name: string): number {
  const result = Number(value);
  if (!Number.isSafeInteger(result) || result < 0) {
    throw new Error(`Invalid ${name}`);
  }
  return result;
}

const wait = (milliseconds: number) =>
  new Promise((resolve) => setTimeout(resolve, milliseconds));

async function main(): Promise<void> {
  actorID = (await request<Actor>("/api/v1/me")).id;
  let delay = 500;
  for (;;) {
    try {
      await connectOnce();
      delay = 500;
    } catch (cause) {
      console.error(cause);
    }
    await wait(delay + Math.random() * 250);
    delay = Math.min(30_000, delay * 2);
  }
}

void main().catch((cause: unknown) => {
  console.error(cause);
  process.exitCode = 1;
});

File: examples/agents/python/agent.py
from __future__ import annotations

import json
import os
import random
import time
import uuid
from typing import Any
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from websockets.sync.client import connect


CORE_URL = os.environ["COMA_CORE_URL"].rstrip("/")
API_KEY = os.environ["COMA_AGENT_API_KEY"].strip()
CONSUMER = os.getenv("COMA_RUNTIME_CONSUMER", "external-python-example")
WS_URL = f"{'wss' if CORE_URL.startswith('https:') else 'ws'}://{CORE_URL.split('://', 1)[1]}/api/v1/ws"


def api(path: str, method: str = "GET", body: Any | None = None) -> Any:
    encoded = None if body is None else json.dumps(body).encode()
    headers = {"Authorization": f"Bearer {API_KEY}"}
    if encoded is not None:
        headers["Content-Type"] = "application/json"
    request = Request(f"{CORE_URL}{path}", data=encoded, headers=headers, method=method)
    try:
        with urlopen(request, timeout=30) as response:
            return json.load(response)
    except HTTPError as error:
        details = error.read().decode(errors="replace")
        raise RuntimeError(f"{method} {path}: {error.code} {details}") from error


def save_checkpoint(seq: int) -> None:
    api(
        f"/api/v1/agent-runtime/checkpoints/{CONSUMER}",
        "PUT",
        {"last_event_seq": seq},
    )


def event_uuid(actor_id: str, label: str) -> str:
    return str(uuid.uuid5(uuid.UUID(actor_id), label))


def is_mention(data: Any, actor_id: str) -> bool:
    return (
        isinstance(data, dict)
        and isinstance(data.get("id"), str)
        and isinstance(data.get("chat_id"), str)
        and data.get("actor_id") != actor_id
        and isinstance(data.get("mentioned_actor_ids"), list)
        and actor_id in data["mentioned_actor_ids"]
    )


def reply(message: dict[str, Any], actor_id: str, seq: int) -> None:
    api(
        "/api/v1/agent-tools/reply_in_thread",
        "POST",
        {
            "arguments": {
                "chat_id": message["chat_id"],
                "client_msg_id": event_uuid(actor_id, f"reply:{seq}"),
                "body": f"I received your mention (message `{message['id']}`).",
                "body_format": "markdown",
                "thread_root_id": message.get("thread_root_id") or message["id"],
                "mentioned_actor_ids": [],
                "file_ids": [],
            },
            "correlation_id": event_uuid(actor_id, f"correlation:{seq}"),
            "confirmed": True,
        },
    )


def connect_once(actor_id: str) -> None:
    checkpoint = api(f"/api/v1/agent-runtime/checkpoints/{CONSUMER}")
    with connect(WS_URL, max_size=262_144, open_timeout=10) as socket:
        socket.send(
            json.dumps(
                {
                    "op": "auth",
                    "request_id": str(uuid.uuid4()),
                    "access_token": API_KEY,
                    "last_seq": checkpoint["last_event_seq"],
                }
            )
        )
        for raw in socket:
            frame = json.loads(raw)
            if frame.get("op") == "hello":
                print(f"Connected as {actor_id}; resume is active.")
                continue
            if frame.get("op") == "resync_required":
                save_checkpoint(int(frame["current_seq"]))
                return
            if frame.get("op") == "error":
                print(f"WebSocket error frame: {frame.get('code', 'unknown')}")
                if frame.get("fatal"):
                    return
                continue
            if frame.get("op") != "event":
                continue

            seq = int(frame["seq"])
            message = frame.get("data")
            if frame.get("type") == "message.created" and is_mention(message, actor_id):
                reply(message, actor_id, seq)
                print(f"Replied to mention {message['id']} at event {seq}.")
            save_checkpoint(seq)
            socket.send(json.dumps({"op": "ack", "seq": seq}))


def main() -> None:
    actor_id = api("/api/v1/me")["id"]
    delay = 0.5
    while True:
        try:
            connect_once(actor_id)
            delay = 0.5
        except KeyboardInterrupt:
            return
        except Exception as error:  # A sample runner should reconnect after transient failures.
            print(error)
        time.sleep(delay + random.random() * 0.25)
        delay = min(30.0, delay * 2)


if __name__ == "__main__":
    main()
